File: apps/api/app/schemas/werkstatt_machines.py
"""Pydantic shapes for individually tracked machines.

Kept out of `schemas/werkstatt.py`, which is already past 800 lines — the
Werkstatt modules follow a per-persona file-size discipline and machines are a
domain of their own rather than another article variant.
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal

from pydantic import AfterValidator, BaseModel, ConfigDict, Field

from app.core.time import to_naive_utc

# Inbound timestamps are normalised to naive UTC at the boundary rather than in
# each service. A browser sends `2026-08-11T18:00:00.000Z` (correctly — that is
# what `toISOString()` produces), the columns are naive, and every service
# compares against a naive `utcnow()`; without this the first booking with an
# explicit return time raises a TypeError deep inside `book_unit`.
#
# Output models keep the plain `datetime`: values coming out of the ORM are
# already naive UTC, and re-validating them would be a no-op with a cost.
NaiveUtcDatetime = Annotated[datetime, AfterValidator(to_naive_utc)]


class _OrmBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class MachineComponentOut(_OrmBase):
    """A sub-component, flattened — enough to show it in the parent's list."""

    id: int
    unit_number: str
    article_id: int
    article_name: str | None = None
    status: str
    serial_number: str | None = None
    next_inspection_due_at: datetime | None = None


class MachineOut(_OrmBase):
    id: int
    unit_number: str
    article_id: int
    # Denormalised for the list view so the client does not need the article
    # separately just to render a row.
    article_name: str | None = None
    manufacturer: str | None = None
    parent_unit_id: int | None = None
    serial_number: str | None = None
    status: str

    current_location_id: int | None = None
    current_location_name: str | None = None
    holder_user_id: int | None = None
    holder_name: str | None = None

    booked_from: datetime | None = None
    booked_until: datetime | None = None
    # True when booked_until is in the past and the machine is still out — the
    # single fact the "overdue" filter is built on, computed server-side so
    # every client agrees on what late means.
    is_overdue: bool = False

    inspection_required: bool = False
    inspection_interval_days: int | None = None
    last_inspected_at: datetime | None = None
    next_inspection_due_at: datetime | None = None
    inspection_overdue: bool = False

    purchased_at: datetime | None = None
    notes: str | None = None
    is_archived: bool = False
    created_at: datetime

    components: list[MachineComponentOut] = Field(default_factory=list)


class MachineCreatePayload(BaseModel):
    article_id: int
    serial_number: str | None = Field(default=None, max_length=120)
    parent_unit_id: int | None = None
    current_location_id: int | None = None
    inspection_required: bool | None = None
    inspection_interval_days: int | None = Field(default=None, ge=1, le=3650)
    last_inspected_at: NaiveUtcDatetime | None = None
    purchased_at: NaiveUtcDatetime | None = None
    notes: str | None = None


class MachineUpdatePayload(BaseModel):
    serial_number: str | None = Field(default=None, max_length=120)
    parent_unit_id: int | None = None
    current_location_id: int | None = None
    status: str | None = None
    inspection_required: bool | None = None
    inspection_interval_days: int | None = Field(default=None, ge=1, le=3650)
    notes: str | None = None
    is_archived: bool | None = None


class MachineBookPayload(BaseModel):
    """Hand a machine to a person, a vehicle, or both.

    `for_today` is what the scanner sends: the overwhelmingly common case is
    "I am taking this now and bringing it back tonight", and making that the
    one-tap default is the difference between the log being kept and not.
    """

    holder_user_id: int | None = None
    to_location_id: int | None = None
    booked_from: NaiveUtcDatetime | None = None
    booked_until: NaiveUtcDatetime | None = None
    for_today: bool = False
    notes: str | None = None


class MachineReturnPayload(BaseModel):
    to_location_id: int | None = None
    # Lets the person handing it back say "this came back broken" in the same
    # action, rather than it going on the shelf and failing for the next crew.
    status: str = "verfuegbar"
    notes: str | None = None


class MachineInspectionPayload(BaseModel):
    inspected_at: NaiveUtcDatetime | None = None
    interval_days: int | None = Field(default=None, ge=1, le=3650)
    passed: bool = True
    notes: str | None = None


class MachineLabelPrintOut(BaseModel):
    """Confirmation that a label job left for the printer.

    `printer` is the "host:port" actually used — surfaced so a wrong IP is
    diagnosable from the client instead of being a silent misprint elsewhere.
    """

    unit_number: str
    printer: str


class MachineLabelBatchItem(BaseModel):
    unit_id: int
    # "gross" = one full sheet; "klein" = quarter label, packed 4-per-sheet.
    format: Literal["gross", "klein"] = "klein"


class MachineLabelBatchPayload(BaseModel):
    """The print queue: what the workshop collected before hitting Drucken."""

    items: list[MachineLabelBatchItem] = Field(min_length=1, max_length=40)


class MachineLabelBatchOut(BaseModel):
    sheets: int
    labels: int
    printer: str


class MachineLabelCapabilitiesOut(BaseModel):
    """What the ACTIVE printer material can carry — drives honest buttons."""

    material_id: str
    material_name: str
    tier: str
    gross: bool
    klein: bool = True
    hint: str | None = None


class MachineMovementOut(_OrmBase):
    """One line of a machine's custody log."""

    id: int
    movement_type: str
    from_location_id: int | None = None
    from_location_name: str | None = None
    to_location_id: int | None = None
    to_location_name: str | None = None
    user_id: int
    user_name: str | None = None
    assignee_user_id: int | None = None
    assignee_name: str | None = None
    expected_return_at: datetime | None = None
    notes: str | None = None
    created_at: datetime

File: apps/web/src/types/werkstattMachines.ts
/**
 * Maschinen (individually tracked machines) — the FE half of the contract with
 * `apps/api/app/schemas/werkstatt_machines.py`.
 *
 * Kept out of `types/werkstatt.ts` for the same reason the backend keeps them
 * out of `schemas/werkstatt.py`: that file is the shared article/stock contract
 * and is already past 800 lines. A machine is not another article variant — it
 * is one physical object with a custody log — so it gets its own contract file.
 *
 * Every field here mirrors a Pydantic field one-for-one. If you change one
 * side, change the other in the same commit.
 */

/**
 * Machine lifecycle. German values on purpose: they are stored in the database
 * and printed in the UI, and translating at the boundary would mean two names
 * for the same state in logs, payloads and screens.
 *
 * Only `verfuegbar` is bookable — see BOOKABLE_STATUSES in
 * `services/werkstatt_machines.py`.
 */
export type MachineStatus =
  | "verfuegbar"
  | "ausgegeben"
  | "wartung"
  | "defekt"
  | "ausgemustert";

/** A sub-component (battery, charger, case) flattened onto its parent. */
export interface MachineComponent {
  id: number;
  unit_number: string;
  article_id: number;
  article_name: string | null;
  status: MachineStatus;
  serial_number: string | null;
  next_inspection_due_at: string | null;
}

export interface Machine {
  id: number;
  /** The label stuck on the machine — "M-0001". Unique and never recycled. */
  unit_number: string;
  article_id: number;
  article_name: string | null;
  manufacturer: string | null;
  parent_unit_id: number | null;
  serial_number: string | null;
  status: MachineStatus;

  current_location_id: number | null;
  current_location_name: string | null;
  holder_user_id: number | null;
  holder_name: string | null;

  booked_from: string | null;
  booked_until: string | null;
  /** Server-computed so "late" means the same thing here and in the scanner. */
  is_overdue: boolean;

  inspection_required: boolean;
  inspection_interval_days: number | null;
  last_inspected_at: string | null;
  next_inspection_due_at: string | null;
  inspection_overdue: boolean;

  purchased_at: string | null;
  notes: string | null;
  is_archived: boolean;
  created_at: string;

  /** Populated by GET /machines/{id} only — the list view leaves it empty. */
  components: MachineComponent[];
}

/** One line of the custody log. */
export interface MachineMovement {
  id: number;
  movement_type: string;
  from_location_id: number | null;
  from_location_name: string | null;
  to_location_id: number | null;
  to_location_name: string | null;
  user_id: number;
  user_name: string | null;
  assignee_user_id: number | null;
  assignee_name: string | null;
  expected_return_at: string | null;
  notes: string | null;
  created_at: string;
}

export interface MachineCreatePayload {
  article_id: number;
  serial_number?: string | null;
  parent_unit_id?: number | null;
  current_location_id?: number | null;
  inspection_required?: boolean | null;
  inspection_interval_days?: number | null;
  last_inspected_at?: string | null;
  purchased_at?: string | null;
  notes?: string | null;
}

export interface MachineUpdatePayload {
  serial_number?: string | null;
  parent_unit_id?: number | null;
  current_location_id?: number | null;
  status?: MachineStatus;
  inspection_required?: boolean;
  inspection_interval_days?: number | null;
  notes?: string | null;
  is_archived?: boolean;
}

export interface MachineBookPayload {
  holder_user_id?: number | null;
  to_location_id?: number | null;
  booked_from?: string | null;
  booked_until?: string | null;
  /** What the scanner sends: taking it now, back tonight. */
  for_today?: boolean;
  notes?: string | null;
}

export interface MachineReturnPayload {
  to_location_id?: number | null;
  /** Lets the person handing it back say "this came back broken". */
  status?: MachineStatus;
  notes?: string | null;
}

export interface MachineInspectionPayload {
  inspected_at?: string | null;
  interval_days?: number | null;
  passed?: boolean;
  notes?: string | null;
}

/** Confirmation from POST /machines/{id}/print-label. */
export interface MachineLabelPrintResult {
  unit_number: string;
  /** "host:port" the job went to — useful when a wrong IP needs diagnosing. */
  printer: string;
}

/** "gross" = one full 99×44 sheet; "klein" = quarter label, 4 per sheet. */
export type MachineLabelFormat = "gross" | "klein";

export interface MachineLabelBatchItem {
  unit_id: number;
  format: MachineLabelFormat;
}

/** The print queue — klein entries may be DIFFERENT machines on one sheet. */
export interface MachineLabelBatchPayload {
  items: MachineLabelBatchItem[];
}

export interface MachineLabelBatchResult {
  sheets: number;
  labels: number;
  printer: string;
}

/** Server-side filters. Mirrors the query params of GET /werkstatt/machines. */
export interface MachineListFilters {
  q?: string;
  status?: MachineStatus;
  location_id?: number;
  holder_user_id?: number;
  overdue_only?: boolean;
  inspection_due_only?: boolean;
  include_archived?: boolean;
  include_components?: boolean;
}
